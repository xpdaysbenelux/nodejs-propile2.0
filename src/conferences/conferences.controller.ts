import {
    Controller,
    UseGuards,
    Post,
    Body,
    Get,
    Param,
    Query,
    Put,
} from '@nestjs/common';
import { ApiTags } from '@nestjs/swagger';

import { RequiredPermissions, UserSession } from '../_shared/decorators';
import {
    RequiredPermissionsGuard,
    AuthenticatedGuard,
} from '../_shared/guards';
import { ConferencesService } from './conferences.service';
import {
    CreateConferenceRequest,
    ConferenceResponse,
    ConferenceIdParam,
    GetConferencesRequestQuery,
    GetConferencesResponse,
    UpdateConferenceRequest,
} from './dto';
import { IUserSession } from '../_shared/constants';
import { ConferencesQueries } from './conferences.queries';

@UseGuards(AuthenticatedGuard)
@Controller('conferences')
@ApiTags('conferences')
export class ConferencesController {
    constructor(
        private readonly conferencesService: ConferencesService,
        private readonly conferenceQueries: ConferencesQueries,
    ) {}

    @RequiredPermissions({ conferences: { view: true, edit: true } })
    @UseGuards(RequiredPermissionsGuard)
    @Get(':conferenceId')
    getConference(
        @Param() params: ConferenceIdParam,
    ): Promise<ConferenceResponse> {
        return this.conferenceQueries.getConference(params.conferenceId);
    }

    @RequiredPermissions({ conferences: { view: true, edit: true } })
    @UseGuards(RequiredPermissionsGuard)
    @Get()
    getConferences(
        @Query() query: GetConferencesRequestQuery,
    ): Promise<GetConferencesResponse> {
        return this.conferenceQueries.getConferences(query);
    }

    @RequiredPermissions({ conferences: { edit: true } })
    @UseGuards(RequiredPermissionsGuard)
    @Post()
    async createConference(
        @Body() body: CreateConferenceRequest,
        @UserSession() session: IUserSession,
    ): Promise<ConferenceResponse> {
        const conferenceId = await this.conferencesService.createConference(
            body,
            session,
        );
        return this.conferenceQueries.getConference(conferenceId);
    }

    @RequiredPermissions({ conferences: { edit: true } })
    @UseGuards(RequiredPermissionsGuard)
    @Put(':conferenceId')
    async updateSession(
        @Body() body: UpdateConferenceRequest,
        @Param() params: ConferenceIdParam,
        @UserSession() userSession: IUserSession,
    ): Promise<string> {
        const conferenceId = await this.conferencesService.updateConference(
            body,
            params.conferenceId,
            userSession,
        );
        return conferenceId;
    }
}
