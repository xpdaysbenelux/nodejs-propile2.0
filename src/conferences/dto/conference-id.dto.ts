import { IsUUID } from 'class-validator';
import { ApiProperty } from '@nestjs/swagger';

export class ConferenceIdParam {
    @ApiProperty({ required: true })
    @IsUUID('4')
    readonly conferenceId: string;
}
