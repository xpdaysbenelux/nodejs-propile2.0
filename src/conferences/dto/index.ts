export * from './create-conference.dto';
export * from './get-conferences.dto';
export * from './get-conference.dto';
export * from './update-conference.dto';
export * from './conference-id.dto';
export * from './room.dto';
