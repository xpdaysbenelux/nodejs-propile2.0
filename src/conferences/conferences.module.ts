import { Module } from '@nestjs/common';

import { ConferencesController } from './conferences.controller';
import { ConferencesService } from './conferences.service';
import { ConferencesQueries } from './conferences.queries';
import { DatabaseModule } from '../database';

@Module({
    imports: [DatabaseModule],
    controllers: [ConferencesController],
    providers: [ConferencesService, ConferencesQueries],
})
export class ConferencesModule {}
