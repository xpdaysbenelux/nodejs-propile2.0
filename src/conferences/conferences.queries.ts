import { Injectable } from '@nestjs/common';
import { ConferenceRepository } from 'src/database';
import { ConferenceResponse } from './dto';

const conferenceFields = [
    'conference.id',
    'conference.name',
    'conference.startDate',
    'conference.endDate',
    'rooms.id',
    'rooms.name',
    'rooms.maxParticipants',
];

@Injectable()
export class ConferencesQueries {
    constructor(private readonly conferenceRepository: ConferenceRepository) {}

    async getConference(conferenceId: string): Promise<ConferenceResponse> {
        return this.conferenceRepository
            .createQueryBuilder('conference')
            .select(conferenceFields)
            .where('conference.id = :conferenceId', { conferenceId })
            .innerJoin('conference.rooms', 'rooms')
            .getOne();
    }
}
