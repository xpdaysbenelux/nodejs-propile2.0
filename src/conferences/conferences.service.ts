import { Injectable } from '@nestjs/common';
import { parseISO } from 'date-fns';

import { ConferenceRepository, Conference, Room } from '../database';
import { CreateConferenceRequest, RoomRequest } from './dto';
import { IUserSession } from '../_shared/constants';
import {
    ConferenceNameAlreadyInUse,
    ConferenceMustHaveAtLeastTwoRooms,
} from './errors';

@Injectable()
export class ConferencesService {
    constructor(private readonly conferenceRepository: ConferenceRepository) {}

    async createConference(
        body: CreateConferenceRequest,
        session: IUserSession,
    ): Promise<string> {
        const { name, startDate, endDate, rooms } = body;

        // Check for unique name
        const existingConference = await this.conferenceRepository.findOne({
            name,
        });
        if (existingConference) {
            throw new ConferenceNameAlreadyInUse();
        }

        const conference = new Conference();
        conference.name = name;
        conference.startDate = new Date(parseISO(startDate));
        conference.endDate = new Date(parseISO(endDate));
        conference.createdBy = session.email;
        conference.createdAt = new Date();
        conference.rooms = this.makeConferenceRooms(rooms);

        await this.conferenceRepository.save(conference);

        return conference.id;
    }

    private makeConferenceRooms(givenRooms: RoomRequest[]): Room[] {
        if (givenRooms.length < 2) {
            throw new ConferenceMustHaveAtLeastTwoRooms();
        }

        const rooms: Room[] = [];
        givenRooms.forEach(room => {
            const newRoom = new Room();
            newRoom.name = room.name;
            newRoom.maxParticipants = room.maxParticipants;
            newRoom.location = room.location || null;

            rooms.push(newRoom);
        });
        return rooms;
    }
}
