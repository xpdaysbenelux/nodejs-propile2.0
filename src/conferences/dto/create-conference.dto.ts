import { IsString, IsDate, ArrayNotEmpty, IsNotEmpty } from 'class-validator';
import { RoomRequest } from './room.dto';

export class CreateConferenceRequest {
    @IsString()
    @IsNotEmpty()
    readonly name: string;

    @IsString()
    @IsNotEmpty()
    readonly startDate: string;

    @IsString()
    @IsNotEmpty()
    readonly endDate: string;

    @ArrayNotEmpty()
    readonly rooms: RoomRequest[];
}
