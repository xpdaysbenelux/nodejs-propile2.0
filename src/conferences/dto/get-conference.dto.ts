import { IsUUID } from 'class-validator';

import { BaseEntityResponse } from '../../_shared/dto';
import { RoomResponse } from '../../rooms/dto/get-room.dto';

export class ConferenceResponse extends BaseEntityResponse {
    readonly name: string;
    readonly startDate: Date;
    readonly endDate: Date;
    readonly rooms: RoomResponse[];
}

export class ConferenceIdParam {
    @IsUUID('4')
    readonly conferenceId: string;
}
