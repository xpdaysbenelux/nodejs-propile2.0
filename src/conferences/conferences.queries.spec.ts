import { Test } from '@nestjs/testing';
import * as faker from 'faker';

import { SortDirection } from '../_shared/constants';
import { ConferenceRepository, DatabaseModule } from '../database';
import { ConferencesQueries } from './conferences.queries';
import { GetConferencesRequestQuery, ConferencesSortColumns } from './dto';

describe('ConferenceQueries', () => {
    let conferenceRepository: ConferenceRepository;
    let conferenceQueries: ConferencesQueries;

    beforeAll(async () => {
        const module = await Test.createTestingModule({
            imports: [DatabaseModule],
            providers: [ConferencesQueries],
        }).compile();

        conferenceQueries = module.get(ConferencesQueries);
        conferenceRepository = module.get(ConferenceRepository);
    });

    afterAll(() => {
        conferenceRepository.manager.connection.close();
    });

    describe('getConference', () => {
        it('should return the requested conference correctly', async () => {
            const result = await conferenceQueries.getConference(
                '81bdc928-b92d-4111-91f9-3ef93e203e88',
            );

            const { startDate, endDate, ...withoutDates } = result;
            expect(withoutDates).toMatchSnapshot();
        });

        it('should return nothing if the requested conference does not exist', async () => {
            const result = await conferenceQueries.getConference(
                faker.random.uuid(),
            );

            expect(result).toMatchSnapshot();
        });
    });

    describe('getConferences', () => {
        const testQueries: GetConferencesRequestQuery[] = [
            {},
            { search: 'Conference' },
            { search: '2' },
            {
                sortBy: ConferencesSortColumns.Name,
                sortDirection: SortDirection.Ascending,
            },
        ];

        testQueries.forEach((query, index) => {
            it(`should return a paged list of conferences: #${index}`, async () => {
                const result = await conferenceQueries.getConferences(query);

                const withoutDatesList = result.data.map(entry => {
                    const { startDate, endDate, ...withoutDates } = entry;
                    return withoutDates;
                });
                expect({
                    data: withoutDatesList,
                    meta: result.meta,
                }).toMatchSnapshot();
            });
        });

        it('should return an empty list if no conferences are found', async () => {
            const result = await conferenceQueries.getConferences({
                search: 'jibbrishsearch',
            });
            expect(result).toMatchSnapshot();
        });
    });
});
