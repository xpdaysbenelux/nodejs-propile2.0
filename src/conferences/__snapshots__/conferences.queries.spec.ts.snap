// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConferenceQueries getConference should return nothing if the requested conference does not exist 1`] = `undefined`;

exports[`ConferenceQueries getConference should return the requested conference correctly 1`] = `
Object {
  "id": "81bdc928-b92d-4111-91f9-3ef93e203e88",
  "name": "Conference 1",
  "rooms": Array [
    Room {
      "id": "8c4bb08a-befe-460c-818a-71c47ecb0761",
      "maxParticipants": 50,
      "name": "Room 1",
    },
    Room {
      "id": "ce63d11d-c51f-4320-a997-fb31b42a6e08",
      "maxParticipants": 50,
      "name": "Room 2",
    },
  ],
}
`;

exports[`ConferenceQueries getConferences should return a paged list of conferences: #0 1`] = `
Object {
  "data": Array [
    Object {
      "id": "3103e833-90d5-47a9-aace-66432881019e",
      "name": "Conference 2",
      "rooms": Array [
        Room {
          "id": "c6ddd134-3edd-4daa-9f59-f29a95650cc3",
          "maxParticipants": 25,
          "name": "Room 7",
        },
        Room {
          "id": "f89d9f18-6611-4520-8bff-8347c70a0b45",
          "maxParticipants": 35,
          "name": "Room 8",
        },
      ],
    },
    Object {
      "id": "81bdc928-b92d-4111-91f9-3ef93e203e88",
      "name": "Conference 1",
      "rooms": Array [
        Room {
          "id": "8c4bb08a-befe-460c-818a-71c47ecb0761",
          "maxParticipants": 50,
          "name": "Room 1",
        },
        Room {
          "id": "ce63d11d-c51f-4320-a997-fb31b42a6e08",
          "maxParticipants": 50,
          "name": "Room 2",
        },
      ],
    },
  ],
  "meta": Object {
    "count": 2,
    "skip": 0,
    "totalCount": 2,
  },
}
`;

exports[`ConferenceQueries getConferences should return a paged list of conferences: #1 1`] = `
Object {
  "data": Array [
    Object {
      "id": "3103e833-90d5-47a9-aace-66432881019e",
      "name": "Conference 2",
      "rooms": Array [
        Room {
          "id": "c6ddd134-3edd-4daa-9f59-f29a95650cc3",
          "maxParticipants": 25,
          "name": "Room 7",
        },
        Room {
          "id": "f89d9f18-6611-4520-8bff-8347c70a0b45",
          "maxParticipants": 35,
          "name": "Room 8",
        },
      ],
    },
    Object {
      "id": "81bdc928-b92d-4111-91f9-3ef93e203e88",
      "name": "Conference 1",
      "rooms": Array [
        Room {
          "id": "8c4bb08a-befe-460c-818a-71c47ecb0761",
          "maxParticipants": 50,
          "name": "Room 1",
        },
        Room {
          "id": "ce63d11d-c51f-4320-a997-fb31b42a6e08",
          "maxParticipants": 50,
          "name": "Room 2",
        },
      ],
    },
  ],
  "meta": Object {
    "count": 2,
    "skip": 0,
    "totalCount": 2,
  },
}
`;

exports[`ConferenceQueries getConferences should return a paged list of conferences: #2 1`] = `
Object {
  "data": Array [
    Object {
      "id": "3103e833-90d5-47a9-aace-66432881019e",
      "name": "Conference 2",
      "rooms": Array [
        Room {
          "id": "c6ddd134-3edd-4daa-9f59-f29a95650cc3",
          "maxParticipants": 25,
          "name": "Room 7",
        },
        Room {
          "id": "f89d9f18-6611-4520-8bff-8347c70a0b45",
          "maxParticipants": 35,
          "name": "Room 8",
        },
      ],
    },
  ],
  "meta": Object {
    "count": 1,
    "skip": 0,
    "totalCount": 1,
  },
}
`;

exports[`ConferenceQueries getConferences should return a paged list of conferences: #3 1`] = `
Object {
  "data": Array [
    Object {
      "id": "81bdc928-b92d-4111-91f9-3ef93e203e88",
      "name": "Conference 1",
      "rooms": Array [
        Room {
          "id": "8c4bb08a-befe-460c-818a-71c47ecb0761",
          "maxParticipants": 50,
          "name": "Room 1",
        },
        Room {
          "id": "ce63d11d-c51f-4320-a997-fb31b42a6e08",
          "maxParticipants": 50,
          "name": "Room 2",
        },
      ],
    },
    Object {
      "id": "3103e833-90d5-47a9-aace-66432881019e",
      "name": "Conference 2",
      "rooms": Array [
        Room {
          "id": "c6ddd134-3edd-4daa-9f59-f29a95650cc3",
          "maxParticipants": 25,
          "name": "Room 7",
        },
        Room {
          "id": "f89d9f18-6611-4520-8bff-8347c70a0b45",
          "maxParticipants": 35,
          "name": "Room 8",
        },
      ],
    },
  ],
  "meta": Object {
    "count": 2,
    "skip": 0,
    "totalCount": 2,
  },
}
`;

exports[`ConferenceQueries getConferences should return an empty list if no conferences are found 1`] = `
Object {
  "data": Array [],
  "meta": Object {
    "count": 0,
    "skip": 0,
    "totalCount": 0,
  },
}
`;
